### PR TITLE
#2719 #2752 #2810 GH Actions for CI Pipeline

### PR DESCRIPTION
**Pull request by christian-kreuzberger-dtx**
This PR addresses the following issues:
- #2719
- #2752 
- #2810

Additional Changes:
* I've introduced a `.ci_env` file which is used to store all CI relevant environment variables
* * I've cleaned certain bash scripts that were used for travis-ci before
* I've created a single "CI Workflow" that handles all the three issues linekd above, as well as unit tests

image

Regarding the *Prepare* job: This checks the repo for changes and other stuff, and prepares environment variables for other jobs, namely:

* `BUILD_EVERYTHING` - indicates that everything should be built (cli, docker images, helm chart)
* `BUILD_{DIRNAME}` - indicates that only a certain directory should be built (e.g., for PRs where only the api or the cli was touched)
* Several metadata like `BRANCH`, `GIT_SHA`, `DATETIME` 
* One important thing: We autogenerate `VERSION` based on
  * If it is a push on a release-branch (e.g., `release-x.y.z`), then `VERSION=x.y.z` 
  * Else, we determine the last "tag" and increase the patch level, e.g., from 0.7.3 to 0.7.4
    * If it is a Pull Request, then `VERSION=${NEXTTAG}-dev-PR-{PRID}`
    * If it is a push on master, then `VERSION=${NEXTTAG}-dev`

**Note 1**: Due to a problem with Unit tests for CLI on Windows I had to disable them
**Note 2**: Due to maintenance of MacOS runners I had to disable everything related to MacOS for now https://github.com/actions/virtual-environments/issues/2247

Docker Images are automatically pushed to DockerHub.
Everything else (CLI build, Helm Chart) is currently only pushed as an artifact of the workflow.

I'll create follow-up issues:
1. Automatically create a `dev` (or `latest`) release which stores the CLI and Helm-Chart such that it can be directly downloaded
2. Automatically create a release draft for changes on release branches
3. Enable MacOS builds again once maintenance is complete
4. Move certain "bash scripts" into a separate re-usable (for keptn-contrib/sandbox) GH action

Signed-off-by: Christian Kreuzberger <christian.kreuzberger@dynatrace.com>

### CODE REVIEW

File: gh-actions-scripts/build_docker_image.sh [hunk 1]
```diff
@@ -1,21 +1,29 @@
...
 cp MANIFEST ./${FOLDER}MANIFEST #$FOLDER contains / at the end
 cp travis-scripts/entrypoint.sh ./${FOLDER}entrypoint.sh #$FOLDER contains / at the end
+
 cd ./${FOLDER}
 
 # uncomment certain lines from Dockerfile that are for Travis builds only
 sed -i '/#travis-uncomment/s/^#travis-uncomment //g' Dockerfile
 cat MANIFEST
-docker build . -t "${IMAGE}:${GIT_SHA}" -t "${IMAGE}:${VERSION}.${DATE}" -t "${IMAGE}:${VERSION}" --build-arg version=$VERSION || travis_terminate 1
+docker build . -t "${IMAGE}:${GIT_SHA}" -t "${IMAGE}:${VERSION}.${DATETIME}" -t "${IMAGE}:${VERSION}" --build-arg version="${VERSION}"
```
Comment (christian-kreuzberger-dtx) [inline gh-actions-scripts/build_docker_image.sh:19]: From @johannes-b : Remove `"${IMAGE}:${GIT_SHA}"`

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+  prepare_ci_run:
+    name: Prepare CI Run
+    # Prepare CI Run looks at what has been changed in this commit/PR/... and determines which artifacts should be
+    # built afterwards (in other jobs that depend on this one).
+    runs-on: ubuntu-20.04
+    outputs: # declare what this job outputs (so it can be re-used for other jobs)
+      # build config
+      BUILD_EVERYTHING: ${{ steps.BUILD_EVERYTHING.outputs.BUILD_EVERYTHING }}
+      BUILD_INSTALLER: ${{ steps.BUILD_EVERYTHING.outputs.BUILD_INSTALLER }}
+      BUILD_API: ${{ steps.BUILD_EVERYTHING.outputs.BUILD_API }}
```
Comment (christian-kreuzberger-dtx) [inline .github/workflows/CI.yml:26]: shoudln't be `BUILD_EVERYTHING`

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+        run: go test -race -v -coverprofile=coverage.txt -covermode=atomic ./...
+        working-directory: ./helm-service
+
+      - name: Test gatekeeper svc
+        if: (needs.prepare_ci_run.outputs.BUILD_EVERYTHING == 'true') || (needs.prepare_ci_run.outputs.BUILD_GATEKEEPER == 'true')
+        run: go test -race -v -coverprofile=coverage.txt -covermode=atomic ./...
+        working-directory: ./gatekeeper-service
+
+      - name: Test distributor svc
+        run: go test -v -coverprofile=coverage.txt -covermode=atomic ./...
```
Comment (christian-kreuzberger-dtx) [inline .github/workflows/CI.yml:302]: missing `if`

File: gh-actions-scripts/build_docker_image.sh (hunk 1 above)
Comment (christian-kreuzberger-dtx) [reply gh-actions-scripts/build_docker_image.sh]: Fixed.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (christian-kreuzberger-dtx) [reply .github/workflows/CI.yml:303]: Fixed.

File: .github/workflows/CI.yml (hunk 2 above)
Comment (christian-kreuzberger-dtx) [reply .github/workflows/CI.yml]: Fixed.

**Review by johannes-b — APPROVED** [review]
lgtm

**Pull request merged**